src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.CoreRT.cs

```diff
@@ -331,9 +364,41 @@ public object GetOrRegisterObjectForComInstance(IntPtr externalComObject, Create
             if (flags.HasFlag(CreateObjectFlags.Aggregation))
                 throw new NotImplementedException();
 
-            object? wrapperMaybeLocal = wrapperMaybe;
-            retValue = null;
-            throw new NotImplementedException();
+            if (!flags.HasFlag(CreateObjectFlags.UniqueInstance))
+            {
+                var comInterfaceDispatch = impl.TryGetComInterfaceDispatch(externalComObject);
+                if (comInterfaceDispatch != null)
+                {
+                    retValue = ComInterfaceDispatch.GetInstance<object>(comInterfaceDispatch);
+                    return true;
+                }
+            }
+
+            retValue = impl.CreateObject(externalComObject, flags);
+            if (retValue == null)
+            {
+                // If ComWrappers instance cannot create wrapper, we can do nothing here.
+                return false;
+            }
+
+            if (flags.HasFlag(CreateObjectFlags.UniqueInstance))
+            {
+                // No need to cache NativeObjectWrapper for unique instances. They are not cached.
+                return true;
+            }
+
+            NativeObjectWrapper wrapper;
+            if (impl._rcwTable.TryGetValue(retValue, out wrapper))
```

Comment:
`impl.CreateObject` above is meant to always create new object. It means that this lookup will never succeed - if it is new object, it could not have been seen before.

Reply:
Again, returning back to responsibility of runtime and implementor of ComWrappers. In the docs https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.comwrappers.createobject?view=net-5.0 `Creates a managed object for the object that externalComObject points to, respecting the values of flags.`. What we do is `not respecting`, i.e. just return always new instances. Maybe wording should be changed somehow? or expanded. I always trying to fall into trap that caching is responsibility of implementor (even if I know that's not true.

Reply:
Also from here https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.comwrappers?view=net-5.0 : `Unlike the built-in features, a majority of the responsibility for accurate lifetime management, dispatching methods, and marshalling of arguments and return values is left to the ComWrappers implementer.` Note  **lifetime management** which for me indicate that all caching/uniquess/addRef should be done by implementer. `The +1 reference count on the native object is assumed to be performed by the managed proxy creator (that is, the ComWrappers implementer) to ensure the associated lifetime between the native object and its managed proxy` from same article, just almost at the end of document.

Reply:
> Unlike the built-in features, a majority of the responsibility for accurate lifetime management, dispatching methods, and marshalling of arguments and return values is left to the ComWrappers implementer.

The next paragraph after that says: 
```
Efficient mapping between a native IUnknown and its managed proxy
```
This mapping is missing in this PR.

> Note lifetime management

It says "a majority of the responsibility for accurate lifetime management". The minority that it is reponsible for is the mapping.

Feel free to submit a PR to the doc repo.

Reply:
This is very delicate topic, and I see myself doing too much errors for now, so I probing my understanding in discussion with you.  I still seek for hidden assumptions, that's why I'm asking same question in different shapes.